src/v/storage/kvstore.cc

```diff
@@ -575,18 +575,28 @@ ss::future<> kvstore::replay_segments(segment_set segs) {
         co_await (*it)->close();
     }
 
+    // we want to defer local storage effects until we've confirmed that
+    // the logical version is compatible with the rest of the cluster
+    _startup_garbage.reserve(std::distance(match, segs.end()) * 2);
+    for (auto it = match; it != segs.end(); it++) {
+        auto seg = *it;
+        _startup_garbage.emplace_back(seg->reader().path().string());
+        _startup_garbage.emplace_back(seg->index().path().string());
+    }
+}
+ss::future<> kvstore::do_startup_compaction() {
     // saving a snapshot right after recovery during start-up prevents an
     // accumulation of segments in cases where the system restarts many
     // times without ever filling up a segment and snapshotting when
     // rolling. they'll be removed on the next startup.
     co_await save_snapshot();
```

Comment:
is it the case that avoiding `save_snapshot` call until after the logical version check is done is key to the solution because if we were to take a snapshot then segments in the `startup_garbage` wouldn't be replayed on the next startup even if they weren't physically remove from disk?

if that's true, then i wonder if there is also a risk of having save_snapshot called via the background flusher fiber. it seems like between `kvstore::start` and `kvstore::do_startup_compaction` that there is a risk that that could occur.

do we need to take additional steps to for that case, or is it not an issue?

Reply:
>if we were to take a snapshot then segments in the startup_garbage wouldn't be replayed on the next startup

Looks that way to me.

>there is also a risk of having save_snapshot called via the background flusher

Yeah, I see what you mean. What's not clear to me is whether this is a real problem; it's not breaking anything today afaik, but it seems like it could. My (somewhat limited) understanding of the terrain in a [comment](https://github.com/redpanda-data/core-internal/issues/739#issuecomment-1708775982) on the issue.

Reply:
We probably only need one example of where it could potentially happen that we should be concerned about it.

Reply:
To be clear, what I mean is that I don't know whether this PR is itself addressing a real problem. If it's worth doing in at all, I agree that we should address the background flusher snapshot as well.

Reply:
got it. i suspect we can probably answer that question by looking at the type of data that is going into the kvstore and trying to reason about if it is sensitive to upgrade/rollback. it seems plausible that there may be some data like that.

Reply:
> is it the case that avoiding `save_snapshot` call until after the logical version check is done is key to the solution because if we were to take a snapshot then segments in the `startup_garbage` wouldn't be replayed on the next startup (even if they weren't physically remove from disk)?


Perhaps the versioned serde should only ever speak a version up to the negotiated current version of the cluster.

We don't wire it up in that way at the moment, but if we do support rollback from a failed upgrade, that might be the best place to do it.

Reply:
That's an interesting idea.

Newbie ahead, sorry: Part of the problem at hand is that the negotiated active version is not yet known when the storage API is starting up, so to implement what you're describing, would we completely block serde until we're past the version check? Is that feasible?

Reply:
> That's an interesting idea.
> 
> Newbie ahead, sorry: Part of the problem at hand is that the negotiated active version is not yet known when the storage API is starting up, so to implement what you're describing, would we completely block serde until we're past the version check? Is that feasible?

We'd probably have to store it somewhere... which makes it hard to bootstrap! I think that the recent changes to serde to use `tag_invoke` provides a mechanism for passing along state such as the current version, so it may be possible to wire it all up from some kind of registry without touching all of the code.